tests/zfs-tests/tests/functional/cli_root/zfs_receive/zfs_receive_016_pos.ksh

```diff
@@ -0,0 +1,85 @@
+#!/bin/ksh -p
+#
+# CDDL HEADER START
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+# CDDL HEADER END
+#
+# Copyright (c) 2020 by Mariusz Zaborski <oshogbo@FreeBSD.org>.
+
+#
+# DESCRIPTION:
+# Verify 'zfs recv' can forcibly unmount filesystem while receiving
+# stream.
+#
+# STRATEGY:
+# 1. Create snapshot of file system
+# 2. Make a zfs filesystem mountpoint busy
+# 3. Recife filesystem with force flag.
+# 4. Verify that stream was received or failed on Linux.
+#
+
+. $STF_SUITE/tests/functional/cli_root/cli_common.kshlib
+
+verify_runnable "both"
+
+function cleanup
+{
+	cd $curpath
+
+	for snap in $init_snap $rst_snap; do
+                snapexists $snap && \
+                        log_must zfs destroy -f $snap
+        done
+
+	datasetexists $rst_root && \
+		log_must zfs destroy -Rf $rst_root
+
+	for file in $full_bkup
+	do
+		[[ -e $file ]] && \
+			log_must rm -f $file
+	done
+
+	[[ -d $TESTDIR1 ]] && \
+		log_must rm -rf $TESTDIR1
```

Comment:
Hym I'm not sure. The purpose is not to know what exactly went wrong?

Reply:
If something fails in cleanup we still want to finish cleaning up. There will be error messages in the logs to know what went wrong.

The purpose of cleanup is to put the system back the way it was when the test started. If there is something that should be tested that is relevant to the test, it can happen in the test rather than in cleanup. Cleanup should make a best effort to clean up as much as possible. It is potentially being invoked halfway through the test because something already went wrong.